src/json-ld.net/Core/DocumentLoader.cs

```diff
@@ -41,189 +83,27 @@ public virtual RemoteDocument LoadDocument(string url)
                     }
                     string header = linkedContexts.First();
                     string linkedUrl = header.Substring(1, header.IndexOf(">") - 1);
-                    string resolvedUrl = URL.Resolve(url, linkedUrl);
+                    string resolvedUrl = URL.Resolve(redirectedUrl, linkedUrl);
                     var remoteContext = this.LoadDocument(resolvedUrl);
                     doc.contextUrl = remoteContext.documentUrl;
                     doc.context = remoteContext.document;
                 }
 
-                Stream stream = resp.GetResponseStream();
+                Stream stream = await httpResponseMessage.Content.ReadAsStreamAsync();
 
-                doc.DocumentUrl = req.Address.ToString();
+                doc.DocumentUrl = redirectedUrl;
                 doc.Document = JSONUtils.FromInputStream(stream);
             }
             catch (JsonLdError)
             {
                 throw;
             }
-            catch (WebException webException)
-            {
-                try
-                {
-                    resp = (HttpWebResponse)webException.Response;
-                    int baseStatusCode = (int)(Math.Floor((double)resp.StatusCode / 100)) * 100;
-                    if (baseStatusCode == 300)
-                    {
-                        string location = resp.Headers[HttpResponseHeader.Location];
-                        if (!string.IsNullOrWhiteSpace(location))
-                        {
-                            // TODO: Add recursion break or simply switch to HttpClient so we don't have to recurse on HTTP redirects.
-                            return LoadDocument(location);
-                        }
-                    }
-                }
-                catch (Exception innerException)
-                {
-                    throw new JsonLdError(JsonLdError.Error.LoadingDocumentFailed, url, innerException);
-                }
-
-                throw new JsonLdError(JsonLdError.Error.LoadingDocumentFailed, url, webException);
-            }
```

Comment:
The semantics here are now covered in the `do {} while()` and the couple of `if`s downfile from it.